libnetwork/resolvconf/resolvconf.go

```diff
@@ -58,32 +66,478 @@ func Path() string {
+		Search         string
+		Options        string
+		Other          string
+		Overrides      []string
+		Comments       bool
+	}{
+		Md:       &rc.md,
+		Search:   strings.Join(rc.search, " "),
+		Options:  strings.Join(rc.options, " "),
+		Other:    rc.other,
+		Comments: comments,
+	}
+	for _, addr := range rc.nameServers {
+		s.NameServers = append(s.NameServers, addr.String())
+	}
+	for _, ns := range rc.md.ExtNameServers {
+		var nss string
+		if ns.HostLoopback {
+			nss = "host(" + ns.Addr.String() + ")"
+		} else {
+			nss = ns.Addr.String()
+		}
+		s.ExtNameServers = append(s.ExtNameServers, nss)
+	}
+	if rc.md.NSOverride {
+		s.Overrides = append(s.Overrides, "nameservers")
+	}
+	if rc.md.SearchOverride {
+		s.Overrides = append(s.Overrides, "search")
+	}
+	if rc.md.OptionsOverride {
+		s.Overrides = append(s.Overrides, "options")
+	}
+
+	templText := `{{if .Comments}}{{with .Md.HeaderComment}}{{.}}
+
+{{end}}{{end}}{{range .NameServers -}}
+nameserver {{.}}
+{{end}}{{with .Search -}}
+search {{.}}
+{{end}}{{with .Options -}}
+options {{.}}
+{{end}}{{with .Other -}}
+{{.}}
+{{end}}{{if .Comments}}
+# Based on host file: '{{.Md.SourcePath}}'{{with .Md.Transform}} ({{.}}){{end}}
+{{if .Md.UsedDefaultNS -}}
+# Used default nameservers.
+{{end -}}
+{{with .ExtNameServers -}}
+# ExtServers: {{.}}
+{{end -}}
+# Overrides: {{.Overrides}}
+{{with .Md.NDotsFrom -}}
+# Option ndots from: {{.}}
+{{end -}}
+{{end -}}
+`
+	var buf bytes.Buffer
+	templ, err := template.New("summary").Parse(templText)
+	if err != nil {
+		return nil, errdefs.System(err)
+	}
+	if err := templ.Execute(&buf, s); err != nil {
+		return nil, errdefs.System(err)
+	}
+	return buf.Bytes(), nil
+}
+
+// WriteFile generates content and writes it to path. If hashPath is non-zero, it
+// also writes a file containing a hash of the content, to enable UserModified()
+// to determine whether the file has been modified.
+func (rc *ResolvConf) WriteFile(path, hashPath string, perm os.FileMode) error {
+	content, err := rc.Generate(true)
+	if err != nil {
+		return err
+	}
+
+	// Write the resolv.conf file - it's bind-mounted into the container, so can't
+	// move a temp file into place, just have to truncate and write it.
+	if err := os.WriteFile(path, content, perm); err != nil {
+		return errdefs.System(err)
+	}
+
+	// Write the hash file.
+	if hashPath != "" {
+		hashFile, err := ioutils.NewAtomicFileWriter(hashPath, perm)
+		if err != nil {
+			return errdefs.System(err)
+		}
+		defer hashFile.Close()
+
+		_, err = hashFile.Write(hashData(content))
+		if err != nil {
+			return errdefs.System(err)
+		}
+	}
+
+	return nil
+}
+
+// UserModified can be used to determine whether the resolv.conf file has been
+// modified since it was generated. It returns true if the files at rcPath and
+// rcHashPath both exist, and the hash file does not contain the hash of the
+// current rc file - else false.
+func UserModified(rcPath, rcHashPath string) bool {
+	currRC, err := os.ReadFile(rcPath)
```

Comment:
I think changing the hash would break live-restore?

On live-restore, we could check whether the hash file is an old one, check whether it matches the current `resolv.conf`, and regenerate the hash if it does ... but that'd mean keeping the old hash function around anyway. I don't think it's worth the complexity?

Reply:
The hash format won't change, though. `hashData` produces an identical digest string

Reply:
Oh ok - thank you ... I'll change it.